File: CoherenceCommons/java/src/com/seovic/coherence/io/pof/EnumPofSerializer.java
/*
Copyright 2009 Aleksandar Seovic

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package com.seovic.coherence.io.pof;


import com.tangosol.io.pof.PofSerializer;
import com.tangosol.io.pof.PofWriter;
import com.tangosol.io.pof.PofReader;
import com.tangosol.io.pof.PofContext;

import java.io.IOException;


/**
 * POF serializer that can be used to serialize all enum values.
 *
 * @author Aleksandar Seovic  2008.10.24
 */
@SuppressWarnings("unchecked")
public class EnumPofSerializer
        implements PofSerializer
    {
    // ---- PofSerializer implementation ------------------------------------

    /**
     * {@inheritDoc}
     */
    public void serialize(PofWriter writer, Object o)
            throws IOException
        {
        if (!o.getClass().isEnum())
            {
            throw new IllegalArgumentException(
                    "EnumPofSerializer can only be used to serialize enum types.");
            }

        writer.writeString(0, ((Enum) o).name());
        writer.writeRemainder(null);
        }

    /**
     * {@inheritDoc}
     */
    public Object deserialize(PofReader reader)
            throws IOException
        {

        PofContext pofContext = reader.getPofContext();
        Class      enumType   = pofContext.getClass(reader.getUserTypeId());

        if (!enumType.isEnum())
            {
            throw new IllegalArgumentException(
                    "EnumPofSerializer can only be used to deserialize enum types.");
            }

        Enum enumValue = Enum.valueOf(enumType, reader.readString(0));
        reader.readRemainder();

        return enumValue;
        }
    }

File: main.net/src/Coherence.Tools/IO/Pof/EnumPofSerializer.cs
/*
Copyright 2009 Aleksandar Seovic

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

using System;
using System.IO;

using Tangosol.IO.Pof;

namespace Seovic.Coherence.IO.Pof
{
    /// <summary><see cref="IPofSerializer"/> implementation that supports
    /// serialization and deserialization of enum values.
    /// </summary>
    /// <author>Aleksandar Seovic  2008.10.30</author>
    public class EnumPofSerializer : IPofSerializer
    {
        #region IPofSerializer implementation

        /// <summary>
        /// Serialize enum value into a POF stream. 
        /// </summary>
        /// <remarks>
        /// This method will simply write the name of an enum value as a
        /// string into a POF stream, in order to achieve maximum portability
        /// across platforms.
        /// </remarks>
        /// <param name="writer">
        /// The <b>IPofWriter</b> with which to write the object's state.
        /// </param>
        /// <param name="o">
        /// The enum value to serialize.
        /// </param>
        /// <exception cref="IOException">
        /// If an I/O error occurs.
        /// </exception>
        public void Serialize(IPofWriter writer, object o)
        {
            if (o == null || !o.GetType().IsEnum)
            {
                throw new ArgumentException(
                        "EnumPofSerializer can only be used to serialize enum types.");
            }

            writer.WriteString(0, o.ToString());
            writer.WriteRemainder(null);
        }

        /// <summary>
        /// Deerialize enum value from a POF stream. 
        /// </summary>
        /// <remarks>
        /// This method expects to find the name of an enum value in the
        /// POF stream.
        /// </remarks>
        /// <param name="reader">
        /// The <b>IPofReader</b> with which to read the object's state.
        /// </param>
        /// <returns>
        /// The deserialized enum instance.
        /// </returns>
        /// <exception cref="IOException">
        /// If an I/O error occurs.
        /// </exception>
        public object Deserialize(IPofReader reader)
        {
            IPofContext pofContext = reader.PofContext;
            Type enumType = pofContext.GetType(reader.UserTypeId);
            if (!enumType.IsEnum)
            {
                throw new ArgumentException(
                        "EnumPofSerializer can only be used to deserialize enum types.");
            }

            object enumValue = Enum.Parse(enumType, reader.ReadString(0));
            reader.ReadRemainder();

            return enumValue;
        }

        #endregion
    }
}